optuna/storages/rdb/storage.py

```diff
@@ -629,3 +682,40 @@ def __del__(self):
 
         if hasattr(self, 'scoped_session'):
             self.remove_session()
+
+
+class _FinishedTrialsCache(object):
+    def __init__(self, enabled):
+        # type: (bool) -> None
+
+        self._finished_trials = {}  # type: Dict[int, structs.FrozenTrial]
+        self._enabled = enabled
+        self._lock = threading.Lock()
+
+    def is_empty(self):
+        # type: () -> bool
+
+        if not self._enabled:
+            return True
+
+        with self._lock:
+            return len(self._finished_trials) == 0
+
+    def cache_trial_if_finished(self, trial):
+        # type: (structs.FrozenTrial) -> None
+
+        if not self._enabled:
+            return
+
+        if trial.state is not structs.TrialState.RUNNING:
```

Comment:
Thank you for letting me this information.

My thought is, 

> changing the definition in this line (i.e., not caching failed trials),

I want to avoid this because I can not think of a reasonable reason to exclude failed trials from caching.

> changing the definition of `TrialState.is_finished()` (i.e., `datatime_complete` is recorded for failed trials),

I like this option. However I do not know why the current implementation does not record `datatime_complete` when a trial fails. If there is an important reason, I will not insist on this option.

> just leaving them inconsistent.

I think this is not a bad choice. However, if we select this, we might should rename `_FinishedTrialsCache` and its methods for avoiding readers confusion.

Reply:
We agreed with selecting the second choice (i.e., changing the definition of `TrialState.is_finished()`).

Reply:
Fixed: https://github.com/pfnet/optuna/pull/349/commits/df84358043184246fdca27a415cfae2c856c80aa

Reply:
Cool! We can update this line as: `if trial.state.is_finished():`.

Reply:
Fixed: https://github.com/pfnet/optuna/pull/349/commits/a934cc2284a8f519fc710cd3b98b316095a2f93f. 
Thank you for pointing it out.